File: zjctools/src/main/java/com/android/zjctools/appupdate/ZUpdateService.java
package com.android.zjctools.appupdate;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.View;
import android.widget.RemoteViews;

import com.android.zjctools.base.ZConstant;
import com.android.zjctools.interface_function.ZCallback;
import com.android.zjctools.interface_function.ZFunctionManager;
import com.android.zjctools.utils.ZFile;
import com.android.zjctools.utils.ZLog;
import com.android.zjctools.utils.ZToast;
import com.android.zjcutils.R;

import java.io.File;


public class ZUpdateService extends IntentService  {


    private String apkUrl,apkDirectory;
    private int appLogoIconRes;
    private ZNotificationUtils notificationUtils;
    private File downApkFile;
    int progress=0;

    public ZUpdateService() {
        super("UpdateService");
    }

    @Override
    public void onCreate() {
        super.onCreate();
    }


    @Override
    protected void onHandleIntent(Intent intent) {
        if (intent != null) {
            notificationUtils =  ZNotificationUtils.getInstance(getApplicationContext());
            apkUrl = intent.getStringExtra(ZConstant.APK_URL);
            apkDirectory = intent.getStringExtra(ZConstant.APK_DIRECTORY);
            appLogoIconRes=intent.getIntExtra(ZConstant.APP_LOGO,-1);
            createNotification();
            download(apkUrl, apkDirectory);
        }
    }

    private void createNotification() {
        RemoteViews remoteViews = new RemoteViews(getPackageName(), R.layout.zjc_widget_notifacion_layout);
        if(appLogoIconRes!=-1){
            remoteViews.setImageViewResource(R.id.notify_download_iv,appLogoIconRes);
        }
        remoteViews.setViewVisibility(R.id.notify_download_iv, View.VISIBLE);
        remoteViews.setTextViewText(R.id.notify_tv, "正在下载...");
        remoteViews.setProgressBar(R.id.notify_progress_pb, 100, 0, false);
        remoteViews.setTextViewText(R.id.notify_progress_tv, "0%");
        notificationUtils.sendNotification( 0,"", "", remoteViews, null);
    }







    /**
     * @param url          下载连接
     */
    public void download(final String url, final String apkDirectory) {
        ZFile.downloadFileByNetWork(url, apkDirectory, ".apk", new ZCallback<File>() {
            @Override
            public void onSuccess(File file) {
                downApkFile=file;
                sendMessage(1,100);
            }

            @Override
            public void onError(int code, String desc) {
                ZLog.e(desc);
                sendMessage(2,0);
            }

            @Override
            public void onProgress(int progress, String desc) {
                sendMessage(0,progress);
            }
        });
    }




    public void sendMessage(int what, int mprogress) {
        if(what==0&&mprogress==progress){
            return;
        }else if(what==0||what==1){
            progress=mprogress;
            Message msg0 = new Message();
            msg0.what = what;
            msg0.arg1 = progress;
            mHandler.sendMessage(msg0);
        }else if(what==2){
            Message msg0 = new Message();
            msg0.what = what;
            mHandler.sendMessage(msg0);
        }

    }


    private Handler mHandler = new Handler(new Handler.Callback() {
        @Override
        public boolean handleMessage(Message msg) {
            switch (msg.what) {
                case 0:
                    //正在下载中
                    ZLog.e("appProgress:"+msg.arg1);
                    ZFunctionManager.getInstance().invokeOnlyPramFunc(ZConstant.App_UPDATE_PROGRESS,msg.arg1);//下载中更新进度条
                    RemoteViews contentView = notificationUtils.getNotification().contentView;
                    contentView.setTextViewText(R.id.notify_tv, "更新包下载中...");
                    contentView.setProgressBar(R.id.notify_progress_pb, 100, msg.arg1, false);
                    contentView.setTextViewText(R.id.notify_progress_tv, msg.arg1 + "%");
                    // 更新UI
                    notificationUtils.notifyNotifaction(0);
                    break;
                case 1://下载完成
                    notificationUtils.cancelNotification(0);
                    ZFunctionManager.getInstance().invokeOnlyPramFunc(ZConstant.APP_DOWNLOAD_FINISH, downApkFile);
                    break;
                case 2:
                    //下载失败
                    ZFunctionManager.getInstance().invokeOnlyPramFunc(ZConstant.App_UPDATE_PROGRESS,-1);//下载中更新进度条
                    break;
            }
            return true;
        }
    });

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

}

File: zjctools/src/main/java/com/android/zjctools/appupdate/ZUpdateService.kt
package com.android.zjctools.appupdate

import android.app.IntentService
import android.content.Intent
import android.os.Handler
import android.os.Message
import android.view.View
import android.widget.RemoteViews
import com.android.zjctools.base.ZConstant
import com.android.zjctools.interface_function.ZDownloadCallback
import com.android.zjctools.utils.ZFile
import com.android.zjctools.utils.ZLog
import com.android.zjcutils.R
import java.io.File


class ZUpdateService : IntentService("UpdateService") {
    private var notificationUtils: ZNotificationUtils? = null

    private var downApkFile: File? = null
    var progress = 0
    override fun onCreate() {
        super.onCreate()
    }

    override fun onHandleIntent(intent: Intent?) {
        if (intent != null) {
            //是否需要处理Notification
            if(ZUpdateManager.appVersion!!.needShowNotification){
                handleNotification()
            }
            download(ZUpdateManager.appVersion!!.apkUrl, ZUpdateManager.appVersion!!.apkDirectory?:"")
        }
    }


    /**
     * 处理Notification
     */
    private fun handleNotification(){
        notificationUtils = ZNotificationUtils.getInstance(applicationContext)
        createNotification()
    }


    /**
     * 创建Notification
     */
    private fun createNotification() {
        val remoteViews = RemoteViews(packageName, R.layout.z_widget_notifacion_layout)
        if (ZUpdateManager.appVersion!!.appLogoIcon != -1) {
            remoteViews.setImageViewResource(R.id.notify_download_iv, ZUpdateManager.appVersion!!.appLogoIcon)
        }
        remoteViews.setViewVisibility(R.id.notify_download_iv, View.VISIBLE)
        remoteViews.setTextViewText(R.id.notify_tv, "正在下载...")
        remoteViews.setProgressBar(R.id.notify_progress_pb, 100, 0, false)
        remoteViews.setTextViewText(R.id.notify_progress_tv, "0%")
        notificationUtils?.sendNotification(0, "", "", remoteViews, null)
    }


    /**
     * 更新通知栏进度条
     */
    private fun showNotificationProgress(process: Int){
        notificationUtils?.let {
            val contentView = notificationUtils!!.getNotification()!!.contentView
            contentView.setTextViewText(R.id.notify_tv, "更新包下载中...")
            contentView.setProgressBar(R.id.notify_progress_pb, 100, process, false)
            contentView.setTextViewText(R.id.notify_progress_tv,  "${process}%")
        }
    }
    /**
     * @param url          下载连接
     */
   private fun download(url: String?="", apkDirectory: String) {
        ZFile.downloadFileByNetWork(url?:"", apkDirectory, ".apk", object : ZDownloadCallback<File> {
            override fun onSuccess(file: File) {
                file?.let {
                    downApkFile = file
                    sendMessage(1, 100)
                }
            }
            override fun onError(code: Int, desc: String?) {
                sendMessage(2, 0)
            }

            override fun onProgress(progress: Int, desc: String?) {
                sendMessage(0, progress)
            }
        })
    }

   private fun sendMessage(what: Int, mprogress: Int) {
        if (what == 0 && mprogress == progress) {
            return
        } else if (what == 0 || what == 1) {
            progress = mprogress
            val msg0 = Message()
            msg0.what = what
            msg0.arg1 = progress
            mHandler.sendMessage(msg0)
        } else if (what == 2) {
            val msg0 = Message()
            msg0.what = what
            mHandler.sendMessage(msg0)
        }
    }

    private val mHandler = Handler { msg ->
        when (msg.what) {
            0 -> {
                //正在下载中
                ZLog.e("appProgress:" + msg.arg1)
                ZUpdateManager.onProgressProgress(msg.arg1)
                showNotificationProgress(msg.arg1)
                // 更新UI
                notificationUtils?.notifyNotifaction(0)
            }
            1 -> { //下载成功
                notificationUtils?.cancelNotification(0)
                downApkFile?.let {
                    ZUpdateManager.onSuccess(it)
                }
            }
            2 -> {   //下载失败
                ZUpdateManager.onFail(-1)
                notificationUtils?.cancelNotification(0)
            }

        }
        true
    }

    override fun onDestroy() {
        super.onDestroy()
    }
}